### PR TITLE
Refine subscribable memory events

### PR DESCRIPTION
**Pull request by eskatos**
This PR address comments made to https://github.com/gradle/gradle/pull/978 after it has been merged.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/initialization/DefaultGradleLauncherFactory.java [hunk 1]
```diff
@@ -63,10 +64,13 @@
     private DefaultGradleLauncher rootBuild;
 
     public DefaultGradleLauncherFactory(
-        ListenerManager listenerManager, ProgressLoggerFactory progressLoggerFactory, GradleUserHomeScopeServiceRegistry userHomeDirServiceRegistry) {
+        ListenerManager listenerManager, MemoryStatusBroadcaster memoryStatusBroadcaster, ProgressLoggerFactory progressLoggerFactory, GradleUserHomeScopeServiceRegistry userHomeDirServiceRegistry) {
         this.globalListenerManager = listenerManager;
         this.userHomeDirServiceRegistry = userHomeDirServiceRegistry;
 
+        // Start broadcasting memory status updates
+        memoryStatusBroadcaster.start();
```
Comment (adammurdoch) [inline subprojects/core/src/main/java/org/gradle/initialization/DefaultGradleLauncherFactory.java]: Should put this somewhere more specific to memory monitoring. I'd just have `MemoryStatusBroadcaster` take care of starting itself on construction, or have `MemoryManager` start it on construction. Or just merge `MemoryStatusBroadcaster` and `MemoryManager`, I don't see any real point in them being separate.

File: subprojects/core/src/main/java/org/gradle/internal/service/scopes/GlobalScopeServices.java [hunk 2]
```diff
@@ -170,8 +172,12 @@ void configure(ServiceRegistration registration, ClassLoaderRegistry classLoader
         }
     }
 
-    GradleLauncherFactory createGradleLauncherFactory(ListenerManager listenerManager, ProgressLoggerFactory progressLoggerFactory, GradleUserHomeScopeServiceRegistry userHomeScopeServiceRegistry) {
-        return new DefaultGradleLauncherFactory(listenerManager, progressLoggerFactory, userHomeScopeServiceRegistry);
+    StoppableScheduledExecutor createScheduledExecutorService(ExecutorFactory executorFactory) {
```
Comment (adammurdoch) [inline subprojects/core/src/main/java/org/gradle/internal/service/scopes/GlobalScopeServices.java]: This is generally not the pattern we use for executors. Instead each thing that needs to do something asynchronously would use `ExecutorFactory` to create an executor and then would also take care of stopping this executor. This way, a service does not accidentally leave jobs running when it stops. I would use this pattern for scheduled execution as well.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/health/DaemonHealthStats.java [hunk 3]
```diff
@@ -32,20 +34,25 @@
...
-    DaemonHealthStats(DaemonRunningStats runningStats, MemoryInfo memoryStatus, GarbageCollectionInfo gcInfo, GarbageCollectionMonitor gcMonitor) {
+    DaemonHealthStats(DaemonRunningStats runningStats, MemoryManager memoryManager, GarbageCollectionInfo gcInfo, GarbageCollectionMonitor gcMonitor) {
         this.runningStats = runningStats;
-        this.memoryStatus = memoryStatus;
         this.gcInfo = gcInfo;
         this.gcMonitor = gcMonitor;
+        memoryManager.addListener(new JvmMemoryStatusListener() {
+            @Override
+            public void onJvmMemoryStatus(JvmMemoryStatus jvmMemoryStatus) {
+                memoryStatus = jvmMemoryStatus;
```
Comment (adammurdoch) [inline subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/health/DaemonHealthStats.java]: Possibly need some kind of synchronisation here. The listener is called on a different thread to that using the memory status.

**Comment by eskatos** [comment]
Thanks @adammurdoch for your comments!
They all should be addressed now, could you have another look?

**Pull request merged**